std: Refactor `ascii.zig` and `mem.zig` and add pointer comparison.

tr1ckydev opened the pull request:
- ~Converted many loops into loop expressions wherever possible.~
- Converted few `while` to `for` loops.
- Converted `while` loops with empty bodies to their continue expression as their bodies.
- Added missing pointer comparison checks.
- Added few missing documentation for functions.

wooster0 commented on lib/std/ascii.zig:
```diff
@@ -179,20 +179,12 @@ pub fn isASCII(c: u8) bool {
 
 /// Uppercases the character and returns it as-is if already uppercase or not a letter.
 pub fn toUpper(c: u8) u8 {
-    if (isLower(c)) {
```
I like the documentation and pointer quality changes but I disagree with these kind of changes. I don't think it necessarily makes it easier to read, rather the opposite sometimes. I think it might be better if you follow the existing style and base your changes on that instead of trying to squeeze stuff on as little lines as possible.

tr1ckydev replied:
I have reverted those changes back to their original form.

mlugg commented on lib/std/mem.zig:
```diff
@@ -1013,29 +1006,25 @@ pub fn indexOfScalar(comptime T: type, slice: []const T, value: T) ?usize {
-    while (i != 0) {
-        i -= 1;
+    var i: usize = slice.len - 1;
+    while (i != 0) : (i -= 1) {
```
This change is just incorrect: this loop has different semantics. It will run for one iteration less, skipping 0.

tr1ckydev replied:
Reverted and fixed now.

mlugg commented on lib/std/mem.zig:
```diff
@@ -1111,21 +1096,20 @@ pub fn indexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize
-        if (i == 0) return null;
-    }
+    return while (true) : (i -= 1) {
+        if (eql(T, haystack[i .. i + needle.len], needle)) break i;
```
This isn't the first or last of these, but this kind of change - which this PR seems in large part focused on - seems to have no real reasoning behind it, it's just using language features for the sake of using language features. Using C-style early returns is more readable, easier to optimize, and takes up fewer characters: I see no reason for these changes.

tr1ckydev replied:
> it's just using language features for the sake of using language features.

That was one of my points of changing them but i have reverted them back.

mlugg reviewed (changes requested):
I agree with wooster's perspective: a lot of these changes feel at best neutral, and at worst a loss for readability. Also, at least one of your changes is a bug (there may be more I didn't spot at a glance).

N00byEdge commented:
Some of these I really like, like
```zig
    var i: usize = start_index;
    while (i < slice.len) : (i += 1) {
```
->
```zig
    for (start_index..slice.len) |i| {
```
But for everything else I'm not as sold on it :(

mlugg commented on lib/std/mem.zig:
```diff
@@ -959,15 +958,14 @@ test "len" {
-    while (ptr[i] != sentinel) {
-        i += 1;
-    }
+    while (ptr[i] != sentinel) : (i += 1) {}
```
What's the point of this change? There's nothing hugely *wrong* with writing it like this, but again, it just feels like using language features for the sake of using them. Every while loop has a body, but not every while loop has a post-expression, so the null hypothesis is to not use a post-expression unless it improves the code.

tr1ckydev replied:
Thanks for the review.
For example,
```zig
pub fn trimLeft(comptime T: type, slice: []const T, values_to_strip: []const T) []const T {
    var begin: usize = 0;
    while (begin < slice.len and indexOfScalar(T, values_to_strip, slice[begin]) != null) : (begin += 1) {}
    return slice[begin..];
}
```
This function has already been there with empty while loop body, so adhering to that I changed that. So should i revert it?

mlugg replied:
Hm, interesting. Personally, yes, I'd change the other use site to use the loop body rather than a post-expression. Post-expressions are effectively syntax sugar for common looping patterns, and when you don't need to use that sugar it just seems like an unnecessary complication. Also, note that nothing stops you from putting the loop on one line, as in `while (cond) i += 1;`.

tr1ckydev replied:
Thats cool, I sticked with that and converted all the empty while loop bodies with their continue exp as their bodies.

mlugg commented on lib/std/mem.zig:
```diff
@@ -3441,13 +3439,14 @@ test "indexOfMinMax" {
     try testing.expectEqual(indexOfMinMax(u8, "a"), .{ .index_min = 0, .index_max = 0 });
 }
 
+/// Swaps the two values.
```
This doc comment is near useless. While this function is very self-documenting anyway, if you're going to add one, at least use clear language (i.e. referencing the actual parameters), e.g.: "Swaps the value pointed to by `a` with the value pointed to by `b`."

tr1ckydev replied:
Thanks, will change that.

IntegratedQuantum commented on lib/std/ascii.zig:
```diff
@@ -328,9 +328,10 @@ test "allocUpperString" {
 
 /// Compares strings `a` and `b` case-insensitively and returns whether they are equal.
 pub fn eqlIgnoreCase(a: []const u8, b: []const u8) bool {
+    if (a.ptr == b.ptr) return true;
```
I think this should placed after the `len` check.
The slices may point at the same memory, but have different lengths.
Consider for example `eqlIgnoreCase("abc", "abc"[0..1])`

tr1ckydev replied:
Woah, never thought about that. Thanks for pointing out. Fixed now.

IntegratedQuantum commented on lib/std/mem.zig:
```diff
@@ -643,8 +642,8 @@ test "lessThan" {
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
-    if (a.len != b.len) return false;
     if (a.ptr == b.ptr) return true;
+    if (a.len != b.len) return false;
```
Same here

tr1ckydev replied:
Done.

andrewrk commented:
If you have the word "Optimized" in your PR title then it should come with a performance delta.

tr1ckydev commented:
> If you have the word "Optimized" in your PR title then the writeup should include performance measurements, before and after.

@andrewrk Apologies, changed the title now. My initial commit, was refactoring most loops to loop expressions which I thought would improve performance but later got to know that there isn't any. So I reverted those changes but the rest of the changes/bug fixes are intact.

andrewrk commented:
This is a chaotic changeset. Since you're a first-time contributor, please focus on making an isolated change with a clear goal. If you want a suggestion for what to work on, have a look at the "contributor friendly" label. To be honest, I'm not really interested in "refactoring" commits from anyone except regular contributors.
